core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1501,13 +1501,16 @@ class KafkaApis(val requestChannel: RequestChannel,
       return
     }
 
+    def updateErrors(producerId: Long, currentErrors: util.HashMap[TopicPartition, Errors]): Unit = {
+      val previousErrors = errors.putIfAbsent(producerId, currentErrors)
+      if (previousErrors != null)
+        previousErrors synchronized previousErrors.putAll(currentErrors)
+    }
+
     def sendResponseCallback(producerId: Long, result: TransactionResult)(responseStatus: Map[TopicPartition, PartitionResponse]): Unit = {
```

Comment:
I'd suggest renaming this to `maybeSendResponseCallback` and to add a comment explaining that this is potentially called concurrently multiple times before the response is sent. It's different enough, that it's worth documenting.